ultradns-rest/src/main/java/denominator/ultradns/UltraDNSRestResourceRecordSetApi.java

```diff
@@ -0,0 +1,205 @@
+import static denominator.common.Preconditions.checkNotNull;
+import static denominator.common.Util.nextOrNull;
+import static denominator.common.Util.toMap;
+
+final class UltraDNSRestResourceRecordSetApi implements denominator.ResourceRecordSetApi {
+
+  private static final int DEFAULT_TTL = 300;
+  private final UltraDNSRest api;
+  private final String zoneName;
+  private final UltraDNSRestRoundRobinPoolApi roundRobinPoolApi;
+
+  UltraDNSRestResourceRecordSetApi(UltraDNSRest api, String zoneName,
+                                   UltraDNSRestRoundRobinPoolApi roundRobinPoolApi) {
+    this.api = api;
+    this.zoneName = zoneName;
+    this.roundRobinPoolApi = roundRobinPoolApi;
+  }
+
+  @Override
+  public Iterator<ResourceRecordSet<?>> iterator() {
+    Iterator<Record> orderedRecords = api
+            .getResourceRecordsOfZone(zoneName)
+            .buildRecords()
+            .iterator();
+    return new GroupByRecordNameAndTypeCustomIterator(orderedRecords);
+  }
+
+  @Override
+  public Iterator<ResourceRecordSet<?>> iterateByName(String name) {
+    checkNotNull(name, "name");
+    final int ANY_RR_TYPE = 255;
+    Iterator<Record> ordered = api
+            .getResourceRecordsOfDNameByType(zoneName, name, ANY_RR_TYPE)
+            .buildRecords()
+            .iterator();
+    return new GroupByRecordNameAndTypeCustomIterator(ordered);
+  }
+
+  @Override
+  public ResourceRecordSet<?> getByNameAndType(String name, String type) {
+    checkNotNull(name, "name");
+    checkNotNull(type, "type");
+    Iterator<Record> orderedRecords = recordsByNameAndType(name, type)
+            .iterator();
+    return nextOrNull(new GroupByRecordNameAndTypeCustomIterator(
+            orderedRecords));
+  }
+
+  private List<Record> recordsByNameAndType(String name, String type) {
+    checkNotNull(name, "name");
+    checkNotNull(type, "type");
+    int typeValue = checkNotNull(lookup(type), "typeValue for %s", type);
+    List<Record> records = null;
+    try {
+      records = api
+              .getResourceRecordsOfDNameByType(zoneName, name, typeValue)
+              .buildRecords();
+    } catch (UltraDNSRestException e) {
+      if (e.code() == UltraDNSRestException.DATA_NOT_FOUND ||
+          e.code() == UltraDNSRestException.RESOURCE_RECORD_POOL_NOT_FOUND) {
+        records = new ArrayList<Record>();
+      } else {
+        throw e;
+      }
+    }
+    return records;
+  }
+
+  @Override
+  public void put(ResourceRecordSet<?> rrset) {
+    checkNotNull(rrset, "rrset was null");
+    checkArgument(!rrset.records().isEmpty(), "rrset was empty %s", rrset);
+    int ttlToApply = rrset.ttl() != null ? rrset.ttl() : DEFAULT_TTL;
+
+    List<Record> toUpdate = recordsByNameAndType(rrset.name(), rrset.type());
+    List<Map<String, Object>> toCreate = new ArrayList<Map<String, Object>>(rrset.records());
+
+    for (Iterator<Record> shouldUpdate = toUpdate.iterator(); shouldUpdate.hasNext();) {
```

Comment:
Same as above. you don’t need the logic to determine which record to add, update or delete for an existing owner.
Just call the PUT of REST & that will automatically take care of this.

So, the high-level logic for the implementation should look something like below:
- Call GET end-point for the input owner, zone & type combination
- If any rrset is returned:
  - Call PUT end-point to update (you already have all the information in the input parameter). Just use the input parameter to map it to the REST input for PUT
- If no rrset is returned:
  - Call POST end-point to create (you already have all the information in the input parameter). Just use the input parameter to map it to the REST input for POST

Please also manually run an integration test on the existing SOAP module & the suggested REST module code to verify that the end results are same.